File: novaquant/trading/risk_manager.py
from typing import Dict, Any, List

class RiskManager:
    """
    Mandatory Risk Management Engine.
    AI suggestions must never bypass risk management.
    """
    def __init__(
        self,
        max_risk_per_trade_pct: float = 1.5,
        max_position_size_pct: float = 15.0,
        min_risk_reward: float = 2.0,
        max_open_trades: int = 3,
        daily_loss_limit_pct: float = 3.0,
        max_leverage: int = 10,
    ):
        self.max_risk_per_trade_pct = max_risk_per_trade_pct
        self.max_position_size_pct = max_position_size_pct
        self.min_risk_reward = min_risk_reward
        self.max_open_trades = max_open_trades
        self.daily_loss_limit_pct = daily_loss_limit_pct
        self.max_leverage = max_leverage

    def evaluate_risk(
        self,
        symbol: str,
        direction: str,
        entry_price: float,
        atr: float,
        account_balance: float,
        daily_pnl: float,
        active_trades_count: int,
        custom_sl: float = None,
        custom_tp: float = None
    ) -> Dict[str, Any]:
        # Rule 1: Daily Loss Limit
        if daily_pnl < -(account_balance * (self.daily_loss_limit_pct / 100)):
            return {
                "approved": False,
                "reason": "Daily loss limit exceeded. Risk engine halted trading.",
                "position_size": 0.0,
                "stop_loss": 0.0,
                "take_profit": 0.0,
                "risk_reward_ratio": 0.0
            }

        # Rule 2: Max Open Trades
        if active_trades_count >= self.max_open_trades:
            return {
                "approved": False,
                "reason": f"Maximum open trades limit reached ({active_trades_count}/{self.max_open_trades})",
                "position_size": 0.0,
                "stop_loss": 0.0,
                "take_profit": 0.0,
                "risk_reward_ratio": 0.0
            }

        # Calculate Stop Loss & Take Profit
        if direction == "LONG":
            sl = custom_sl or round(entry_price - (1.5 * atr), 2)
            tp = custom_tp or round(entry_price + (3.2 * atr), 2)
        else:
            sl = custom_sl or round(entry_price + (1.5 * atr), 2)
            tp = custom_tp or round(entry_price - (3.2 * atr), 2)

        risk_dist = abs(entry_price - sl)
        reward_dist = abs(tp - entry_price)

        if risk_dist <= 0:
            return {"approved": False, "reason": "Invalid stop loss distance"}

        rr_ratio = round(reward_dist / risk_dist, 2)

        # Rule 3: Minimum 1:2 Risk/Reward Ratio
        if rr_ratio < self.min_risk_reward:
            return {
                "approved": False,
                "reason": f"Risk/Reward ratio {rr_ratio}:1 does not satisfy minimum {self.min_risk_reward}:1",
                "position_size": 0.0,
                "stop_loss": sl,
                "take_profit": tp,
                "risk_reward_ratio": rr_ratio
            }

        # Rule 4: Position Sizing (Fixed fractional risk)
        risk_capital = account_balance * (self.max_risk_per_trade_pct / 100)
        pos_size = round(risk_capital / risk_dist, 3 if entry_price > 1000 else 1)

        return {
            "approved": True,
            "reason": "Trade meets all institutional risk requirements",
            "position_size": pos_size,
            "stop_loss": sl,
            "take_profit": tp,
            "risk_reward_ratio": rr_ratio,
            "risk_amount": round(risk_dist * pos_size, 2),
            "potential_profit": round(reward_dist * pos_size, 2)
        }

File: server/riskManager.ts
import { RiskSettings, TradeDirection, SupportedExchange } from '../src/types';
import { CapitalState } from './capitalManager';

export interface OrderIntent {
  symbol: string;
  direction: TradeDirection;
  aiSuggestedSize?: number;
  aiSuggestedCapital?: number;
  currentPrice: number;
  atr?: number;
  customStopLoss?: number;
  customTakeProfit?: number;
  clientOrderId?: string;
}

export interface RiskValidationResult {
  approved: boolean;
  rejectedReason?: string;
  clampedPositionSize: number;
  allocatedCapital: number;
  stopLoss: number;
  takeProfit: number;
  riskRewardRatio: number;
  riskAmount: number;
  potentialProfit: number;
  warnings: string[];
}

export class RiskManager {
  private recentOrders: Array<{ symbol: string; direction: string; timestamp: number; clientOrderId?: string }> = [];
  private isEmergencyStopActive: boolean = false;

  triggerEmergencyStop(reason: string = 'User or circuit breaker emergency stop triggered'): void {
    this.isEmergencyStopActive = true;
    console.warn(`[EMERGENCY STOP TRIGGERED] Trading halted: ${reason}`);
  }

  resetEmergencyStop(): void {
    this.isEmergencyStopActive = false;
    console.log('[EMERGENCY STOP RESET] Trading operations resumed.');
  }

  isEmergencyStopped(): boolean {
    return this.isEmergencyStopActive;
  }

  /**
   * Enforces all risk rules before any order reaches the Capital Manager or Exchange Adapter:
   * 1. Emergency Stop Check
   * 2. Duplicate Order Protection
   * 3. Max Open Positions Check
   * 4. Daily Loss Limit Check
   * 5. Stop Loss & Take Profit Mandatory Enforcement (R:R Ratio >= minRiskRewardRatio)
   * 6. Capital Allocation Cap & AI Override Protection
   * 7. Insufficient Balance Protection
   */
  evaluateOrderRisk(
    intent: OrderIntent,
    capitalState: CapitalState,
    riskSettings: RiskSettings,
    activeTradeCount: number,
    dailyPnL: number,
    dailyStartingCapital: number
  ): RiskValidationResult {
    const warnings: string[] = [];

    // 1. Emergency Stop Check
    if (this.isEmergencyStopActive) {
      return {
        approved: false,
        rejectedReason: 'Emergency Stop is currently active. All new order submissions are strictly halted.',
        clampedPositionSize: 0,
        allocatedCapital: 0,
        stopLoss: 0,
        takeProfit: 0,
        riskRewardRatio: 0,
        riskAmount: 0,
        potentialProfit: 0,
        warnings: ['Emergency Stop Triggered'],
      };
    }

    // 2. Duplicate Order Protection (Within 15 seconds for same symbol + direction or identical clientOrderId)
    const now = Date.now();
    const DUPLICATE_WINDOW_MS = 15000;
    // Clean old history
    this.recentOrders = this.recentOrders.filter((o) => now - o.timestamp < DUPLICATE_WINDOW_MS);

    const isDuplicate = this.recentOrders.some(
      (o) =>
        (intent.clientOrderId && o.clientOrderId === intent.clientOrderId) ||
        (o.symbol === intent.symbol && o.direction === intent.direction && now - o.timestamp < DUPLICATE_WINDOW_MS)
    );

    if (isDuplicate) {
      return {
        approved: false,
        rejectedReason: `Duplicate Order Protection: A recent order for ${intent.symbol} (${intent.direction}) was placed within the last 15s. Request rejected to prevent double-execution.`,
        clampedPositionSize: 0,
        allocatedCapital: 0,
        stopLoss: 0,
        takeProfit: 0,
        riskRewardRatio: 0,
        riskAmount: 0,
        potentialProfit: 0,
        warnings: ['Duplicate order prevented'],
      };
    }

    // 3. Maximum Number of Open Positions (Strictly 1 trade at a time)
    const maxTradesAllowed = Math.max(1, riskSettings.maxOpenTrades || 1);
    if (activeTradeCount >= maxTradesAllowed) {
      return {
        approved: false,
        rejectedReason: `Strict 1-Trade Rule Active: Position capacity reached (${activeTradeCount}/${maxTradesAllowed}). Only 1 trade can run at a time. The current position must close before the next trade can execute.`,
        clampedPositionSize: 0,
        allocatedCapital: 0,
        stopLoss: 0,
        takeProfit: 0,
        riskRewardRatio: 0,
        riskAmount: 0,
        potentialProfit: 0,
        warnings: ['Single-trade concurrency active (1 trade at a time)'],
      };
    }

    // 4. Maximum Daily Loss Limit
    const startingCap = dailyStartingCapital > 0 ? dailyStartingCapital : capitalState.allocatedCapital;
    const dailyPnLPct = startingCap > 0 ? (dailyPnL / startingCap) * 100 : 0;
    if (dailyPnLPct <= -riskSettings.dailyLossLimitPercent) {
      return {
        approved: false,
        rejectedReason: `Daily loss limit reached (${dailyPnLPct.toFixed(2)}% vs max allowable -${riskSettings.dailyLossLimitPercent}%). Daily circuit breaker active.`,
        clampedPositionSize: 0,
        allocatedCapital: 0,
        stopLoss: 0,
        takeProfit: 0,
        riskRewardRatio: 0,
        riskAmount: 0,
        potentialProfit: 0,
        warnings: ['Daily circuit breaker triggered'],
      };
    }

    // 5. Calculate ATR-based Stop Loss & Take Profit with Mandatory R:R Ratio Check
    const currentPrice = intent.currentPrice > 0 ? intent.currentPrice : 65000;
    const atr = intent.atr || currentPrice * 0.015;

    let stopLoss = intent.customStopLoss;
    let takeProfit = intent.customTakeProfit;

    if (intent.direction === 'LONG') {
      if (!stopLoss) stopLoss = Number((currentPrice - riskSettings.atrMultiplierSL * atr).toFixed(2));
      if (!takeProfit) takeProfit = Number((currentPrice + riskSettings.atrMultiplierTP * atr).toFixed(2));
    } else {
      if (!stopLoss) stopLoss = Number((currentPrice + riskSettings.atrMultiplierSL * atr).toFixed(2));
      if (!takeProfit) takeProfit = Number((currentPrice - riskSettings.atrMultiplierTP * atr).toFixed(2));
    }

    const priceRisk = Math.abs(currentPrice - stopLoss);
    const priceReward = Math.abs(takeProfit - currentPrice);

    if (priceRisk <= 0) {
      return {
        approved: false,
        rejectedReason: 'Invalid Stop Loss: Price risk distance cannot be zero.',
        clampedPositionSize: 0,
        allocatedCapital: 0,
        stopLoss,
        takeProfit,
        riskRewardRatio: 0,
        riskAmount: 0,
        potentialProfit: 0,
        warnings: ['Zero risk distance'],
      };
    }

    const riskRewardRatio = Number((priceReward / priceRisk).toFixed(2));
    if (riskRewardRatio < riskSettings.minRiskRewardRatio) {
      return {
        approved: false,
        rejectedReason: `Risk/Reward ratio (${riskRewardRatio}:1) is below the institutional policy threshold of ${riskSettings.minRiskRewardRatio}:1.`,
        clampedPositionSize: 0,
        allocatedCapital: 0,
        stopLoss,
        takeProfit,
        riskRewardRatio,
        riskAmount: 0,
        potentialProfit: 0,
        warnings: ['Unacceptable R:R ratio'],
      };
    }

    // 6. Mandatory Capital Sizing & AI Override Prevention
    // Rule: Never allow the AI model itself to directly decide how much money it can spend.
    // Example: Available Capital = $1,000, Max Position Size = 10% -> Max order capital = $100.
    const availableTradingCap = capitalState.availableTradingCapital;
    if (availableTradingCap <= 0) {
      return {
        approved: false,
        rejectedReason: 'Insufficient balance protection: Available trading capital is $0.00 USDT.',
        clampedPositionSize: 0,
        allocatedCapital: 0,
        stopLoss,
        takeProfit,
        riskRewardRatio,
        riskAmount: 0,
        potentialProfit: 0,
        warnings: ['Zero trading capital'],
      };
    }

    // Max capital per trade is capped by maxRiskPerTradePercent (or maxPositionSizePercent)
    const maxAllowedCapital = Math.min(
      availableTradingCap,
      availableTradingCap * (riskSettings.maxPositionSizePercent / 100)
    );

    // Max risk dollar amount allowed (e.g. 1.5% of total allocated capital)
    const maxDollarRisk = capitalState.allocatedCapital * (riskSettings.maxRiskPerTradePercent / 100);

    // Position size based on risk amount: size = maxDollarRisk / priceRisk
    const riskSizedUnits = maxDollarRisk / priceRisk;

    // Position size based on max allowable capital: size = maxAllowedCapital / currentPrice
    const capitalSizedUnits = maxAllowedCapital / currentPrice;

    // The final size MUST NOT exceed capital-sized units or risk-sized units
    let finalUnits = Math.min(riskSizedUnits, capitalSizedUnits);

    // If AI proposed a size, clamp it strictly to our maximum allowable size
    if (intent.aiSuggestedSize && intent.aiSuggestedSize > 0) {
      if (intent.aiSuggestedSize > finalUnits) {
        warnings.push(`AI requested ${intent.aiSuggestedSize} units, strictly clamped down to ${finalUnits.toFixed(4)} units by Risk Engine policy.`);
      }
      finalUnits = Math.min(intent.aiSuggestedSize, finalUnits);
    }

    const precision = currentPrice > 1000 ? 4 : 2;
    const finalSize = Math.max(0.0001, Number(finalUnits.toFixed(precision)));
    const requiredCapital = Number((finalSize * currentPrice).toFixed(2));

    // 7. Insufficient balance protection
    if (requiredCapital > availableTradingCap) {
      return {
        approved: false,
        rejectedReason: `Insufficient balance: Required trade margin ($${requiredCapital.toFixed(2)} USDT) exceeds available capital ($${availableTradingCap.toFixed(2)} USDT).`,
        clampedPositionSize: 0,
        allocatedCapital: 0,
        stopLoss,
        takeProfit,
        riskRewardRatio,
        riskAmount: 0,
        potentialProfit: 0,
        warnings: ['Balance exceeded'],
      };
    }

    const actualRisk = Number((priceRisk * finalSize).toFixed(2));
    const actualReward = Number((priceReward * finalSize).toFixed(2));

    // Record order in recent history for duplicate protection
    this.recentOrders.push({
      symbol: intent.symbol,
      direction: intent.direction,
      timestamp: now,
      clientOrderId: intent.clientOrderId,
    });

    return {
      approved: true,
      clampedPositionSize: finalSize,
      allocatedCapital: requiredCapital,
      stopLoss,
      takeProfit,
      riskRewardRatio,
      riskAmount: actualRisk,
      potentialProfit: actualReward,
      warnings,
    };
  }
}

export const riskManager = new RiskManager();
